File: Python/94.Binary Tree Inorder Traversal.py
#写出二叉树的中序遍历
class TreeNode(object):
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution(object):
    def inorderTraversal(self, root):
        self.result = []
        self.inorder(root)
        return self.result

    def inorder(self,p):
        if p!=None:
            self.inorder(p.left) if p.left!=None else None
            self.result.append(p.val)
            self.inorder(p.right) if p.right!=None else None


t1 = TreeNode(1)
t2 = TreeNode(2)
t3 = TreeNode(3)
t4 = TreeNode(4)
t5 = TreeNode(5)
t6 = TreeNode(6)
t1.left = t2
t1.right = t3
t3.left = t4
t4.right = t5

s = Solution()
print(s.inorderTraversal(t1))

File: C++/Hash Table/94.Binary Tree Inorder Traversal.cpp
/*
中序遍历二叉树。
*/
/*
思路：。
*/
#include <iostream>
#include <vector>
#include <stack>
using namespace std;

struct TreeNode {
	int val;
	TreeNode *left;
	TreeNode *right;
	TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution {

	/*递归方法，时间较长。
	public:
	vector<int> result;
	public:
	vector<int> inorderTraversal(TreeNode* root) {
	inorder(root);
	return result;
	}
	void inorder(TreeNode* root)
	{
	if(root)
	{
	inorder(root->left);
	result.push_back(root->val);
	inorder(root->right);
	}
	}*/

public:
	//使用堆栈
	vector<int> inorderTraversal(TreeNode* root) {
		vector<int> result;
		stack<TreeNode*>	mystack;
		TreeNode *cur = root;
		while (!mystack.empty() || cur)
		{
			if (cur)
			{
				mystack.push(cur);
				cur = cur->left;
			}
			else
			{
				cur = mystack.top();
				result.push_back(cur->val);
				mystack.pop();
				cur = cur->right;
			}
		}
		return result;
	}
};

int main()
{
	Solution s;
	TreeNode root(1);
	TreeNode t2(2);
	TreeNode t3(3);
	root.right = &t2;
	t2.left = &t3;
	vector<int> out;
	out = s.inorderTraversal(&root);
	for (int i = 0; i < out.size(); ++i)
	{
		cout << out[i] << endl;
	}
}

